### ISSUE TITLE
Run Ant Media Server in Azure Kubernetes & Create documentation  

### ISSUE THREAD

**Issue by mekya** [ant-media/Ant-Media-Server#3511]

**Comment by mekya** [comment]
Sure @dvuljanic ,

@yashtandon113 , please let us know the updates through this issue as well.

**Comment by muratugureminoglu** [comment]
Hello @dvuljanic , our Kubernetes documentation on Azure will be available soon. But I made an installation for you and I am sharing the steps below. If you have any questions, please don't hesitate to contact me. Also When we publish the document, we will add it under this topic.

Let's create a Kubernetes service on Azure.

**1.** `az login`
**2.** `az group create --location "West Europe" --name aks`
**3.** `az aks create --resource-group aks --name aks-cluster --node-count 1 -s Standard_F8s_v2	 --kubernetes-version 1.20.7`
**4.** `az aks show --name aks-cluster --resource-group aks`
**5.** `az aks get-credentials --name aks-cluster --resource-group aks`
**6.** `kubectl get nodes`
**7.** There are the following yml files that I use, you can change it according to yourself.

#Mongodb
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: mongodb
spec:
  selector:
    matchLabels:
      run: mongodb
  replicas: 1
  template:
    metadata:
      labels:
        run: mongodb
    spec:
      #hostNetwork: true
      containers:
      - name: mongodb
        imagePullPolicy: Always
        image: mongodb:4.2
        ports:
        - containerPort: 27017
```
#Ant Media Server
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ant-media-server
spec:
  selector:
    matchLabels:
      app: ant-media
  replicas: 1
  template:
    metadata:
      labels:
        app: ant-media
    spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: app
                operator: In
                values:
                - ant-media
            topologyKey: "kubernetes.io/hostname"
      hostNetwork: true
      containers:
      - name: ant-media-server
        imagePullPolicy: Always
        image: antmedia-k8s
        args: ["-g", "true", "-s", "true", "-r", "true", "-m", "cluster", "-h", "mongodb_server"]
        resources:
          requests:
             cpu: 1000m
```
#Service
```
kind: Service
apiVersion: v1
metadata:
  name: ant-media-server
spec:
  selector:
    app: ant-media
  ports:
    - name: http
      protocol: TCP
      port: 5080 
```
#HPA > check this document please: https://github.com/ant-media/Ant-Media-Server/wiki/Kubernetes-Autoscaling
```
apiVersion: autoscaling/v1
kind: HorizontalPodAutoscaler
metadata:
  name: ant-media-server
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: ant-media-server
  minReplicas: 1
  maxReplicas: 10
  targetCPUUtilizationPercentage: 60
```
#Ingress > https://docs.microsoft.com/en-us/azure/aks/ingress-basic
```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: ant-media-server
  annotations:
    kubernetes.io/ingress.class: nginx
    nginx.ingress.kubernetes.io/affinity: "cookie"
    nginx.ingress.kubernetes.io/session-cookie-name: "route"
    nginx.ingress.kubernetes.io/session-cookie-expires: "172800"
    nginx.ingress.kubernetes.io/session-cookie-max-age: "172800"
spec:
  rules:
  - host: test.antmedia.xyz
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: ant-media-server
            port:
              number: 5080

  tls:
  - secretName: antmedia-cert
    hosts:
      - yourdomain.com
```
**8.** Import your SSL certificate
```
kubectl create secret tls antmedia-cert --key="ams.key" --cert="ams.crt"
```

**9.** Check Ingress with the command below and then you can access it from the browser.
```
kubectl get ingress
```

**Comment by dvuljanic** [comment]
@muratugureminoglu Thanks for letting me know, this looks quite similar to the approach that we have used (although worked both with and without ingress). One question - did you actually try to perform some streaming on this setup in Azure with more than 1 pod being active (for example by manually setting minReplicas to 2)? This is the piece of the puzzle that makes most problems for us, causing streaming errors.

**Comment by dvuljanic** [comment]
@muratugureminoglu Appreciate clear instructions in here. We did also try this on our end and I can confirm initial tests manage to resolve issues that we had (with a little bit of additional code needed from our side):
- ingress with cookie session affinity resolved AMS dashboard issues (automatically logged out)
- also, this seems as resolving the issue publishing / playing streams in browser
- to make it work on mobile apps (Android), we needed to additionally change websocket library (seems the one we used before was not fully compatible with sticky sessions :()

Seems the missing part in final environment on our end was the ingress with sticky sessions. We will do some more tests on our end and verify stability of the system (with AMS 2.4.0). Once again, appreciate your support on this one.
